[Navigation] Support "initial about:blank document" logic

rwlbuis opened the pull request:
#### 5a53929d85a19d356f2d0c0b40a13e84165db281
<pre>
[Navigation] Support &quot;initial about:blank document&quot; logic
<a href="https://bugs.webkit.org/show_bug.cgi?id=274393">https://bugs.webkit.org/show_bug.cgi?id=274393</a>

Reviewed by Darin Adler.

Keep track of the &quot;initial about:blank document&quot; flag and check it in
the &quot;has entries and events disabled&quot; algorithm.

[1] <a href="https://html.spec.whatwg.org/multipage/nav-history-apis.html#has-entries-and-events-disabled">https://html.spec.whatwg.org/multipage/nav-history-apis.html#has-entries-and-events-disabled</a> (Step 3)

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigate-from-initial-about-blank-same-doc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigate-from-initial-about-blank-same-doc-popup-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/exception-order-initial-about-blank-unserializablestate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/initial-about-blank-expected.txt:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::maybeLoadEmpty):
* Source/WebCore/loader/DocumentLoader.h:
(WebCore::DocumentLoader::isInitialAboutBlank const):
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::hasEntriesAndEventsDisabled const):

Canonical link: <a href="https://commits.webkit.org/280378@main">https://commits.webkit.org/280378@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a94e87cd7b1ff2a1785f540cdd699936f047795b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56456 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35782 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8928 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60063 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6892 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58582 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43404 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7086 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45730 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4822 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58485 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33648 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48720 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26592 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30425 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6045 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/5896 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6317 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61747 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/364 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6431 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52991 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/364 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48781 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/52885 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12498 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/324 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31609 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32695 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33778 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32442 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/dom/Document.h:
```diff
@@ -2597,6 +2600,8 @@ class Document
     bool m_scheduledDeferredAXObjectCacheUpdate { false };
     bool m_wasRemovedLastRefCalled { false };
 
+    bool m_isInitialAboutBlank { false };
```
The FrameLoader already has this state:
bool isShowingInitialAboutBlank = m_localFrame->loader().stateMachine().isDisplayingInitialEmptyDocument()

rwlbuis replied:
At the location where that code is used the information is valid, however the state quickly moves on to FirstLayoutDone, so I feel we still need to store it somewhere so the information remains static (until next navigation). The latest upload tries to do that, PTAL.

darinadler commented on Source/WebCore/loader/FrameLoader.h:
```diff
@@ -536,6 +538,8 @@ class FrameLoader final : public CanMakeCheckedPtr<FrameLoader> {
 
     bool m_errorOccurredInLoading { false };
+
+    bool m_isInitialAboutBlank { true };
```
This seems like a flag that should be on a document loader. Why does the frame loader need to copy the flag out?

rwlbuis replied:
You are right, a flag on DocumentLoader or Document is better, done.

darinadler commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -386,6 +386,8 @@ bool Navigation::hasEntriesAndEventsDisabled() const
     if (!window()->document() || !window()->document()->isFullyActive())
         return true;
+    if (frame() && frame()->loader().isInitialAboutBlank())
+        return true;
```
Do we really need another flag here? Can we get the same result by getting to the document loader and checking `isInFinishedLoadingOfEmptyDocument`?

rwlbuis replied:
Unfortunately isInFinishedLoadingOfEmptyDocument has the same problem as the state machine, this is not static and in fact resets when the end of maybeLoadEmpty is reached. It will be even more important to keep the value static (until next navigation) once we schedule Navigate event dispatching on the event loop, like https://github.com/WebKit/WebKit/pull/28101 intends to do.

rwlbuis committed:
[Navigation] Support "initial about:blank document" logic
https://bugs.webkit.org/show_bug.cgi?id=274393

Reviewed by Darin Adler.

Keep track of the "initial about:blank document" flag and check it in
the "has entries and events disabled" algorithm.

[1] https://html.spec.whatwg.org/multipage/nav-history-apis.html#has-entries-and-events-disabled (Step 3)

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigate-from-initial-about-blank-same-doc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigate-from-initial-about-blank-same-doc-popup-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/exception-order-initial-about-blank-unserializablestate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/initial-about-blank-expected.txt:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::maybeLoadEmpty):
* Source/WebCore/loader/DocumentLoader.h:
(WebCore::DocumentLoader::isInitialAboutBlank const):
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::hasEntriesAndEventsDisabled const):

Canonical link: https://commits.webkit.org/280378@main
